dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileTinyVectorIndexWriter.cpp

```diff
@@ -0,0 +1,234 @@
+        }
+
+        if (any_new_index_build)
+        {
+            build.file_ids.emplace_back(LocalIndexerScheduler::ColumnFileTinyID(tiny_file->getDataPageId()));
+        }
+    }
+
+    build.file_ids.shrink_to_fit();
+    return build;
+}
+
+ColumnFileTinyPtr ColumnFileTinyVectorIndexWriter::buildIndexForFile(
+    const ColumnDefines & column_defines,
+    const ColumnDefine & del_cd,
+    const ColumnFileTiny * file,
+    ProceedCheckFn should_proceed) const
+{
+    // read_columns are: DEL_MARK, COL_A, COL_B, ...
+    // index_builders are: COL_A, COL_B, ...
+
+    ColumnDefinesPtr read_columns = std::make_shared<ColumnDefines>();
+    read_columns->reserve(options.index_infos->size() + 1);
+    read_columns->push_back(del_cd);
+
+    std::unordered_map<ColId, std::vector<VectorIndexBuilderPtr>> index_builders;
+
+    std::unordered_map<ColId, std::vector<LocalIndexInfo>> col_indexes;
+    for (const auto & index_info : *options.index_infos)
+    {
+        if (index_info.type != IndexType::Vector)
+            continue;
+        col_indexes[index_info.column_id].emplace_back(index_info);
+    }
+
+    for (const auto & [col_id, index_infos] : col_indexes)
+    {
+        // Make sure the column_id is in the schema.
+        const auto cd_iter = std::find_if(column_defines.cbegin(), column_defines.cend(), [&](const auto & cd) {
+            return cd.id == col_id;
+        });
+        RUNTIME_CHECK_MSG(
+            cd_iter != column_defines.cend(),
+            "Cannot find column_id={} in file_id={}",
+            col_id,
+            file->getDataPageId());
+
+        for (const auto & idx_info : index_infos)
+        {
+            // Just skip if the index is already built
+            if (file->hasIndex(idx_info.index_id))
+                continue;
+
+            index_builders[col_id].emplace_back(
+                VectorIndexBuilder::create(idx_info.index_id, idx_info.index_definition));
+        }
+        read_columns->push_back(*cd_iter);
+    }
+
+    // If no index to build, return nullptr
+    if (read_columns->size() == 1 || index_builders.empty())
+        return nullptr;
+
+    // Read all blocks and build index
+    const size_t num_cols = read_columns->size();
+    ColumnFileTinyReader reader(*file, options.data_provider, read_columns);
+    while (true)
+    {
+        if (!should_proceed())
+            throw Exception(ErrorCodes::ABORTED, "Index build is interrupted");
+
+        auto block = reader.readNextBlock();
+        if (!block)
+            break;
+
+        RUNTIME_CHECK(block.columns() == read_columns->size());
+        RUNTIME_CHECK(block.getByPosition(0).column_id == TAG_COLUMN_ID);
+
+        auto del_mark_col = block.safeGetByPosition(0).column;
+        RUNTIME_CHECK(del_mark_col != nullptr);
+        const auto * del_mark = static_cast<const ColumnVector<UInt8> *>(del_mark_col.get());
+        RUNTIME_CHECK(del_mark != nullptr);
+
+        for (size_t col_idx = 1; col_idx < num_cols; ++col_idx)
```

Comment:
A possibly better pattern is to read one column and build one index, then read another column and build another index. In this way, the peak memory usage is reduced.

It can be just a `// TODO`, as I believe this is a rare case to optimize.

Reply:
Add a TODO.